docs/proposals/20191030-machine-health-checking.md

```diff
@@ -0,0 +1,221 @@
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [User Stories [optional]](#user-stories-optional)
+      - [Story 1](#story-1)
+      - [Story 2](#story-2)
+    - [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Design Details](#design-details)
+    - [Test Plan](#test-plan)
+    - [Graduation Criteria](#graduation-criteria)
+      - [Examples](#examples)
+        - [Alpha -> Beta Graduation](#alpha---beta-graduation)
+        - [Beta -> GA Graduation](#beta---ga-graduation)
+        - [Removing a deprecated flag](#removing-a-deprecated-flag)
+    - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+    - [Version Skew Strategy](#version-skew-strategy)
+  - [Implementation History](#implementation-history)
+  - [Drawbacks [optional]](#drawbacks-optional)
+  - [Alternatives [optional]](#alternatives-optional)
+  - [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+
+Enable opt in automated health checking and remediation of unhealthy nodes backed by machines.
+
+## Motivation
+
+- Reduce administrative overhead to run a cluster.
+
+- Ability to respond to failures of machines and keep the cluster nodes healthy.
+
+### Goals
+
+- Enable automated remediation for pools of machines/nodes
+
+- Allow to define different unhealthy criterias based on node conditions for different pools.
+
+- Providing administrator-defined short-circuiting of automated remediation when multiple nodes are unhealthy at the same time.
+
+### Non-Goals/Future Work
+
+- Recording long-term stable history of all health-check failures or remediations.
+
+## Proposal
+
+The machine health checker (MHC) does a best effort to keep nodes healthy in the cluster.
+
+It provides a short-circuit mechanism and limits remediation when `maxUnhealthy` threshold is reached for a targeted pool. This is similar to what the node life cycle controller does for reducing the eviction rate as nodes goes unhealthy in a given zone. E.g a large number of nodes in a single zone are down because it's most likely a networking issue.
+
+The machine health checker is an integration point between node problem detection tooling expresed as node conditions and remediation to achieve a node auto repairing feature.
+
+
+### Unhealthy criteria:
+A machine/node target is unhealthy when:
+
+- The node meets the unhealthy node conditions criteria defined.
+- The Machine has no nodeRef.
+- The Machine has nodeRef but node is not found.
+- The Machine is in phase "Failed".
+
+If any of those criterias are met for longer than given timeouts, remediation is triggered.
+For the node conditions the time outs are defined by the admin. For the other cases opinionated values can be assumed.
```

Comment:
Is the admin here a Management Cluster admin or a Workload Cluster admin?

Reply:
Are these opinionated values meant to be hardcoded defaults? Or is this intended to provide some type of a fallback if the user doesn't define those values?

Reply:
I'd assumed hardcoded values unless we think there's a good reason to allow them to be tweakable

Reply:
I'm good with that approach, we can always expose additional configuration of those values as we see fit.